lib/DBIx/TransactionManager/Distributed.pm

```diff
@@ -0,0 +1,268 @@
+    delete $DBH_SOURCE{$category}{$addr};
+    # avoiding grep here because these are weakrefs and we want them to stay that way.
+    # since they're weakrefs, some of these may be undef
+    extract_by { $addr == (defined($_) ? refaddr($_) : 0) } @{$DBH{$category}};
+    return $dbh;
+}
+
+=head2 dbh_is_registered
+
+Returns true if the provided database handle has been registered already.
+
+Used when registering a handle acquired via L<DBI/connect_cached>.
+
+    register_dbh($category => $dbh) unless dbh_is_registered($category => $dbh);
+
+=cut
+
+sub dbh_is_registered {
+    my ($category, $dbh) = @_;
+    die "too many parameters to register_dbh: @_" if @_ > 2;
+    _check_fork();
+    my $addr = refaddr $dbh;
+    return exists $DBH_SOURCE{$category}{$addr} ? 1 : 0;
+}
+
+=head2 register_cached_dbh
+
+Records the given database handle created via L<DBI/connect_cached> as being active and available for running transactions against.
+
+Expects a category (string value) and L<DBI::db> instance.
+
+Returns the database handle.
+
+Example:
+
+    sub _dbh {
+        my $dbh = DBI->connect_cached('dbi:Pg', '', '', { RaiseError => 1});
+        return register_cached_dbh('category' => $dbh);
+    }
+
+=cut
+
+sub register_cached_dbh {
+    my ($category, $dbh) = @_;
+    register_dbh($category => $dbh) unless BOM::Database::dbh_is_registered($category => $dbh);
+    $dbh->begin_work if $IN_TRANSACTION && $dbh->{AutoCommit};
+    $dbh;
+}
+
+=head2 txn
+
+Runs the given coderef in a transaction.
+
+Expects a coderef and one or more database handle categories.
+
+Will call L<DBI/begin_work> for every known database handle in the given category,
+run the code, then call L<DBI/commit> on success, or L<DBI/rollback> on failure.
+
+Will raise an exception on failure, or return an empty list on success.
+
+Example:
+
+    txn { dbh()->do('NOTIFY something') } 'category';
+
+WARNING: This only applies transactions to known database handles. Anything else -
+Redis, cache layers, files on disk - is out of scope. Transactions are a simple
+L<DBI/begin_work> / L<DBI/commit> pair, there's no 2-phase commit or other
+distributed transaction co-ordination happening here.
+
+=cut
+
+sub txn(&;@) {
+    my ($code, @categories) = @_;
+    die "Need a database category" unless @categories;
+    _check_fork();
+    my $wantarray = wantarray;
+    for my $category (@categories) {
+        if (my $count = () = extract_by { !defined($_) } @{$DBH{$category}}) {
+            warn "Had $count database handles that were not released via release_dbh, probable candidates follow:\n";
+            my %addr = map { ; refaddr($_) => 1 } @{$DBH{$category}};
+            warn "unreleased dbh in $_\n" for sort delete @{$DBH_SOURCE{$category}}{grep !exists $addr{$_}, keys %{$DBH_SOURCE{$category}}};
+        }
+    }
+
+    my @rslt;
+    eval {
+        for my $category (@categories) {
+            $_->begin_work for @{$DBH{$category}};
```

Comment:
@chylli We could raise an exception if it's not set - since the flag can be changed at any time, we'd need to check before the `->begin_work` call and again before `->commit`, not sure if it's worth the effort. Could just document that the module expects `RaiseError` to be set - up to you.